### PR TITLE
Zug-compatible seigniorage, part 1

### PR DESCRIPTION
**Pull request by AlexanderLimonov**

### CODE REVIEW

File: node/src/types/block.rs
Comment (afck) [inline node/src/types/block.rs:59]: Why change the example?

I guess we need to discuss whether the `rewards` field should be removed entirely in 2.0?

File: node/src/components/consensus/highway_core/finality_detector/horizon.rs
Comment (afck) [inline node/src/components/consensus/highway_core/finality_detector/horizon.rs:101]: You can remove the method if it was only used in the rewards computation.

File: node/src/components/contract_runtime/operations.rs
Comment (afck) [inline node/src/components/contract_runtime/operations.rs:76]: We should discuss whether we want to store the `ExecutionResult`s that represent this state change.

File: execution_engine/src/core/engine_state/mod.rs
Comment (afck) [inline execution_engine/src/core/engine_state/mod.rs:1823]:
```suggestion
            // seeds address generator with block timestamp
```

File: execution_engine/src/system/auction.rs
Comment (afck) [inline execution_engine/src/system/auction.rs:558]: I guess this should be unreachable anyway, and even if, it would do the right thing (compute 0 rewards) anyway?
I'd prefer to either keep or remove the code, rather than commenting it out.

File: execution_engine/src/system/auction.rs
Comment (afck) [inline execution_engine/src/system/auction.rs:581]: (This is not a new issue, of course, but:)
We should make sure that and document why these errors are unreachable.

I think both `delegator_reward` and `commission` have at most 64 * 2 = 128 resp. 64 * 3 = 192 bits in the denominator and numerator, so after the subtraction the numerator has at most 320, and will then be reduced back to at most 192. (If I understand correctly, the crate automatically reduces fractions?) In `delegator_rewars` another 64 bits are added (⇒ ≤ 256). `reinvest_delegator_rewards` seems to immediately round them to an integer, so that should be fine — would feel a bit safer to round them here and not use `Ratio` in `reinvest_delegator_rewards`.

And I guess the subtraction here can't fail because we don't allow `commission`s greater than 100%? (Do we check that?) It might be safer to just `saturating_sub` here?

File: execution_engine/src/system/auction.rs
Comment (afck) [inline execution_engine/src/system/auction.rs:603]: Is this guaranteed to not underflow, because `to_integer` rounds towards zero, so `total_delegator_payout` will always be less than `total_reward`?

**Review by afck — COMMENTED** [review]
Looks good so far (haven't fully reviewed the auction tests file yet).
Definitely needs another pair of eyes, preferably one more familiar with EE.

File: execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs
Comment (afck) [inline execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs:1526]: So that's just `expected_total_reward`, is it?

File: execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs
Comment (afck) [inline execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs:1530]:
```suggestion
        let delegator_stake_before = U512::from(DELEGATOR_1_STAKE);
```

(Also below.)

**Review by afck — APPROVED** [review]

File: execution_engine_testing/test_support/src/wasm_test_builder.rs
Comment (fizyk20) [inline execution_engine_testing/test_support/src/wasm_test_builder.rs:1364]: Should we still leave an option to add some rewards while advancing eras? Of course, it would need to work differently, so maybe it's not worth the added complexity? (Not sure what it was being used for.)

File: execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs
Comment (fizyk20) [inline execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs:1332]: My personal preference would be to have the expected payouts calculated by hand and hardcoded here - this code seems to be pretty much a copy of the code that is in `distribute_rewards`, so as it is now, we're basically checking the correctness of the code against itself.

File: node/src/types/block.rs
Comment (AlexanderLimonov) [reply node/src/types/block.rs:59]: The rewards struct floating around is now always an empty vector, it's only there to maintain compatibility with the now-independent `casper-client`

File: execution_engine_testing/test_support/src/wasm_test_builder.rs
Comment (AlexanderLimonov) [reply execution_engine_testing/test_support/src/wasm_test_builder.rs:1364]: There's a second part coming (rewards for propagation of finality signatures), but I am yet to work out a design. Note that this is just a function in the test support crate.

File: execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs
Comment (AlexanderLimonov) [reply execution_engine_testing/tests/src/test/system_contracts/auction/distribute.rs:1332]: Nearly all of these tests should be simplified and re-written, but I only fixed them up enough to have them pass.

File: node/src/types/block.rs
Comment (afck) [reply node/src/types/block.rs:59]: Makes sense.
But maybe create a ticket for 2.0.0 to revisit this.

File: execution_engine/src/system/auction.rs
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/system/auction.rs:581]: we can and should reevaluate this, but Alex doesn't need to be blocked on it for this PR

File: execution_engine_testing/test_support/src/wasm_test_builder.rs
Comment (EdHastingsCasperAssociation) [reply execution_engine_testing/test_support/src/wasm_test_builder.rs:1364]: seems to be primarily used for advancing past initial eras pre-populated at genesis in tests

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Comment by AlexanderLimonov** [comment]
bors merge

**Comment by AlexanderLimonov** [comment]
bors retry

**Pull request merged**
